### DISCUSSION TITLE
getSignedUrl not working as expected

### DISCUSSION THREAD

**Discussion by premraghvani** [aws/aws-sdk-js#4328] [Q&A]
Good day all,

I have been getting the "SignatureDoesNotMatch" error recently when trying to get a signed URL for putObject.
My code for putObject is:
```
try {
            const url = await s3.getSignedUrl('putObject', {
                Bucket: config.s3.tables, 
                Key: `image.png`,
                Expires: 900,
                ContentType: 'image/png; charset=UTF-8',
            })
            body = url
            console.log(url)
        } catch(e) {
            console.log(e);
        }
```

I have tried for getObject, replacing the key name and 'putObject' and it works like a charm.

I defined the AWS using:
```
const AWS = require("aws-sdk");
AWS.config.update({region:"eu-west-2"});
const s3 = new AWS.S3({signatureVersion: "v4",region:"eu-west-2"});
```

Here is part of the error:
```
<Code>SignatureDoesNotMatch</Code>
<Message>
The request signature we calculated does not match the signature you provided. Check your key and signing method.
</Message>
```

And for context, this is all running off of a Lambda function in eu-west-2, the bucket is also there. The IAM role allows this Lambda function to both PutObject and GetObject from this bucket (specified in `config.s3.tables`)

Thanks in advance

**Comment by ajredniwja** [comment]
Hey @premraghvani thanks for opening this discussion, please refer to the [API docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getSignedUrl-property). 

You must ensure that you have static or previously resolved credentials if you call this method synchronously (with no callback), otherwise it may not properly sign the request. If you cannot guarantee this (you are using an asynchronous credential provider, i.e., EC2 IAM roles), you should always call this method with an asynchronous callback.

Also see: [getSignedUrlPromise](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getSignedUrlPromise-property). 

Please reach out if you have any other questions.
